x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -651,7 +653,75 @@ public LogicalPlan apply(LogicalPlan logicalPlan, AnalyzerContext context) {
             } else {
                 limit = context.configuration().resultTruncationMaxSize(); // user provided a limit: cap result entries to the max
             }
-            return new Limit(Source.EMPTY, new Literal(Source.EMPTY, limit, DataTypes.INTEGER), logicalPlan);
+
+            Limit l = new Limit(Source.EMPTY, new Literal(Source.EMPTY, limit, DataTypes.INTEGER), logicalPlan);
+            return maybeAddDefaultLimitForMvExpand(l);
+        }
+
+        /**
+         * This adds the implicit limit to a plan that has a sort and no limit between EsRelation and the first MvExpand.
+         * To date, the only known use case that "needs" this is a query of the form
+         * from test
+         * | sort emp_no
+         * | mv_expand first_name
+         * | rename first_name AS x
+         * | where x LIKE "*a*"
+         * | limit 15
+         *
+         * or
+         *
+         * from test
+         * | sort emp_no
+         * | mv_expand first_name
+         * | sort first_name
+         * | limit 15
+         *
+         * LogicalPlanAnalyzer.PushDownAndCombineLimits rule will copy the "limit 15" after "sort emp_no" if there is no filter
+         * on the expanded values OR if there is no sort between "limit" and "mv_expand". In these two situations, to be able for
+         * "sort emp_no" to form a TopN, we need a limit. Since the "limit 15" in the query cannot be pushed down (otherwise, it will change
+         * the results that reach mv_expand command) we need some kind of value there. For now, this is the implicit limit.
+         * The second query above becomes:
+         *
+         * from test
+         * | sort emp_no
+         * | limit 10000
```

Comment:
I'm afraid this doesn't work: adding a limit there changes the query semantics by filtering out candidate records too early. 
in this case, the `| sort emp_no` will be overridden by `| sort first_name`, so it can be completely removed and the query will become 
```
from test
| mv_expand first_name
| sort first_name
| limit 15
```

In general, the problem is not that we cannot push LIMIT past MV_EXPAND; the following transformation is absolutelly valid:
`| mv_expand x | limit 10` -> `| limit 10 | mv_expand x | limit 10`
and also the default limit is valid as long as it's higher than the original limit:
`| mv_expand x | limit 10` -> `| limit 10000 | mv_expand x | limit 10`

The real problem here is that we cannot completely remove the `limit 10` after `mv_expand`, so we end up trying to push it down it again in an infinite loop.
My biggest concern is that even if we find a way to avoid it now, we'll eventually add new rules to the planner and we'll hardly take this infinite loop into account, so it will eventually happen again.

Reply:
> I'm afraid this doesn't work: adding a limit there changes the query semantics by filtering out candidate records too early.

Agree with you, but as the planner works now - where a TopN is needed - a limit is required at that specific place in the query. [As I mentioned in the code comments](https://github.com/elastic/elasticsearch/pull/103080/files#diff-2c689d9c0606d310fd36a603064f43cb4b23d356f08983fbdb98c65e76e6919dR682) **for now** I decided to use the implicit limit.

> In this specific case, the `| sort emp_no` will be overridden by `| sort first_name`

What do you mean "overriden"? The sorts do not change. If you meant "the sort first_name anyway changes the order of the final documents, so it makes sort emp_no obsolete" then this is not true. `sort emp_no` followed by `mv_expand` and then `sort first_name` has to happen in this specific order for things to make sense.

> cannot push LIMIT past MV_EXPAN

I think this is a misunderstanding on your part. The code in the optimizer doesn't push `limit`, but it makes a copy of `limit` and pushes the copy. The `limit` after `mv_expand` doesn't disappear.

Reply:
> What do you mean "overriden"?

I mean that the following operations do not depend on the input order (none of the existing commands do, apart from LIMIT), so the fact that you sort records by `emp_no` doesn't make any difference: data will be re-sorted by `first_name` afterwards anyway

> The code in the optimizer doesn't push limit, but it makes a copy of limit and pushes the copy

conceptually there is no big difference, you are adding a new limit before mv_expand (that can be considered a pushdown) and you are also preserving the original limit.